runtime/commonMain/src/kotlinx/serialization/json/JsonElementSerializer.kt

```diff
@@ -67,8 +67,9 @@ public object JsonPrimitiveSerializer : KSerializer<JsonPrimitive> {
 
     override fun deserialize(decoder: Decoder): JsonPrimitive {
         verify(decoder)
-        return if (decoder.decodeNotNullMark()) JsonPrimitive(decoder.decodeString())
-        else decoder.decodeSerializableValue(JsonNullSerializer)
+        val result = (decoder as JsonInput).decodeJson()
```

Comment:
WDYT of using already existing `.asJsonInput()` method?

Reply:
Good. Though I am not sure whether it should throw SE instead of ISE. It's not "some serialization error", it's an explicit violation of API contract.
Don't mind if I change the type of the exception to ISE?